core/src/main/java/org/infinispan/marshall/core/GlobalMarshaller.java

```diff
@@ -31,124 +39,413 @@
-       * or
-       * - objectFromByteBuffer() // underneath it calls start/finish
-       * So, there's only need to read version during the start.
-       * First option is preferred when multiple objects are gonna be written.
-       */
+   public BufferSizePredictor getBufferSizePredictor(Object o) {
+      return marshallableTypeHints.getBufferSizePredictor(o.getClass());
+   }
+
+   @Override
+   public ByteBuffer objectToBuffer(Object o) throws IOException, InterruptedException {
       try {
-         return defaultMarshaller.objectFromObjectStream(in);
-      } catch (EOFException e) {
-         IOException ee = new EOFException(
-               "The stream ended unexpectedly.  Please check whether the source of " +
-                     "the stream encountered any issues generating the stream.");
-         ee.initCause(e);
-         throw ee;
-      } catch (IOException ioe) {
-         if (trace) log.trace("Log exception reported", ioe);
-         if (ioe.getCause() instanceof InterruptedException)
-            throw (InterruptedException) ioe.getCause();
-         else
-            throw ioe;
+         BytesObjectOutput out = writeObjectOutput(o);
+         return out.toByteBuffer();
+      } catch (java.io.NotSerializableException nse) {
+         if (log.isDebugEnabled()) log.debug("Object is not serializable", nse);
+         throw new NotSerializableException(nse.getMessage(), nse.getCause());
       }
    }
 
    @Override
-   public boolean isMarshallable(Object o) throws Exception {
-      return defaultMarshaller.isMarshallable(o);
+   public byte[] objectToByteBuffer(Object obj, int estimatedSize) throws IOException, InterruptedException {
+      try {
+         BytesObjectOutput out = writeObjectOutput(obj, estimatedSize);
+         return out.toBytes();
+      } catch (java.io.NotSerializableException nse) {
+         if (log.isDebugEnabled()) log.debug("Object is not serializable", nse);
+         throw new NotSerializableException(nse.getMessage(), nse.getCause());
+      }
+   }
+
+   @Override
+   public ObjectInput startObjectInput(InputStream is, boolean isReentrant) {
+      throw new UnsupportedOperationException("No longer in use");
+   }
+
+   @Override
+   public void finishObjectInput(ObjectInput oi) {
+      throw new UnsupportedOperationException("No longer in use");
+   }
+
+   @Override
+   public Object objectFromObjectStream(ObjectInput in) {
+      throw new UnsupportedOperationException("No longer in use");
+   }
+
+   static final int ID_NULL                        = 0x00;
+   static final int ID_PRIMITIVE                   = 0x01;
+   static final int ID_INTERNAL                    = 0x02;
+   static final int ID_EXTERNAL                    = 0x03;
+   static final int ID_ANNOTATED                   = 0x04;
+   static final int ID_UNKNOWN                     = 0x05;
+
+   static final int NOT_FOUND                      = -1;
```

Comment:
Minor, but would it be better to declare these at the top of the class so that they are more visible?